accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -495,10 +502,24 @@ of a function call produces a result of type `Never`.
 
 The static type of a `throw e` expression is `Never`.
 
-In legacy mode, an override of operator== with no explicit parameter type
+Consider an expression `e` of the form `e1 == e2` where the static type of
+`e1` is `T1` and the static type of `e2` is `T2`. Let `S` be the type of the
+formal parameter of `operator ==` in the interface of **NonNull**(`T1`).
+It is a compile-time error unless `T2` is assignable to `S?`.
```

Comment:
Should we state explicitly that the return type of S.operator== is irrelevant and the static type of `e1==e2` is always bool?

Reply:
I think it falls out naturally, but it doesn't hurt to be explicit.  Also, we should specify whether `Never x = throw "error"; var b = x == 3;` infers `Never` for ` b` or `bool`.

Reply:
I'd say that it infers `bool` because the static type of `e1 == e2` is always `bool`. Much simpler that way.

I don't want to special-case `Never`. It's not worth that much effort. Whatever we do should be general and apply to any return type of `operator==` and not assume that it's just `bool` or `Never` and that `Never` is special in known ways.

The run-time behavior is to either return a boolean if either operand is `null`, or return the result of calling `operator==`, which is (currently) restricted to subtypes of `bool`, which currently means `bool` or `Never`.
So, the correct static type would be LUB(`bool`, return-type-of-`operator==`). That's currently `bool`.

I know that `Never` is the only subtype of `bool`, so allowing `x==3` to have any type other than `bool` would mean it's `Never`, and `x == 3` will definitely never evaluate to a value (because `x` won't). But that's a lot of special knowledge.